File: src/pds.c
#include <stdio.h>
#include <stdlib.h>

/* check_pds takes in the flattened array of a square matrix C,
 * its one-sided dimension p, and a tolerance (ideally, say, 10^-9).
 */

double* check_sym(double* C, int p, double tol) {
    double F_C = 0;
    int i,j;
    for (i=0; i < p; i++) {
        for (j=0; j < p; j++) {
            // Get the Froebenius norm of C - t(C).
            F_C += (C[i*p + j] - C[j * p + i]) * (C[i*p + j] - C[j * p + i]);
	}
    }
    if (F_C < tol) {
        // Allocate C_new.
	double* C_new = malloc(sizeof(C));
        
	for (i=0; i<p; i++) for (j=0; j<p; j++) C_new[i*p + j] = (C[i*p + j] + C[j*p + i])/2;
	return C_new;
    } else {
        printf("%dx%d input matrix is not symmetric within tolerance %.12f.\n", p, p, tol);
	exit(1);
    }
}

File: R/pds.R
library(matrixcalc)

#' Checks whether a matrix is positive definite symmetric, within a given tolerance.
#' @param C The matrix to be evaluated
#' @param tol The tolerance to determine if C is symmetric. If C - t(C) is greater than or equal to this tolerance, an error is thrown. Otherwise, C is set to the average of C and t(C).
#' @returns The matrix after forcing its symmetry.
#' @export
check.pds <- function(C, tol = 10^-9) {
  if (norm(C - t(C), type="F") < tol) {
	  C <- (C + t(C))/2
          if (!is.positive.definite(C)) {
		  cat("Warning: Matrix is not positive definite. Listing eigenvalues:", eigen(C)$values, "\n")
	  }
	  return(C)
  } else {
	  throw("Matrix symmetry is outside of tolerance: ",tol)
  }
}

#' Turns the array of the entries of a triangular matrix into a symmetric matrix and returns said matrix.
#' @param M.tri The array of triangular values, including the diagonal entries.
#' @param p The dimension of the full matrix.
#' @param lower Whether M.tri represents the lower- or upper-triangular values. Defaults to FALSE.
#' @returns M, the full symmetric matrix comprised of the entries of M.tri.

tri.to.sym <- function(M.tri, p, lower=FALSE) {
    M <- matrix(0, p, p)
    if (lower) {
        M[lower.tri(M, diag=TRUE)] <- M.tri
        M[upper.tri(M)] <- t(M)[upper.tri(M)]
    } else {
        M[upper.tri(M, diag=TRUE)] <- M.tri
        M[lower.tri(M)] <- t(M)[lower.tri(M)]
    }
    return(M)
}

